### DISCUSSION TITLE
Dagger action configuration practices

### DISCUSSION THREAD

**Discussion by sagikazarmark** [dagger/dagger#2744] [General]
In the past couple months I've created a few Dagger actions, some of them behave a bit differently (by default) when running in a local environment compared to CI or when running them on PRs compared to tags.

Typically, publishing steps should "dry run" by default in a local environment / non-tag CI runs.

So far I used the following style:

```cue
package main

dagger.#Plan & {
	client: filesystem: ".": read: _
	client: env: {
		CI: string | *""
	}
	actions: {
		someAction: myaction.#DoSomething {
			dryRun: client.env.CI != "true"
		}
	}
}
```

Since the `CI` env var is fairly common to be set by CIs, I want to make that a default within `myaction.#DoSomething`.

That however requires to manually pass the CI env var to the action which is not shorter either.

Any plans to provide guidelines for these kind of config options/make default configuration based on the environment easier?

**Comment by marcosnils** [comment]
Related: https://github.com/dagger/dagger/issues/2164

**Comment by helderco** [comment]
Seems to me the current convention is what you're already doing. I.e., do a field comprehension based on an env var's value or something similar.

If you mean that you wish your action had access to the client's env vars without having to pass through from `client`, you're not the first to ask (#2437).

If not, do you have an idea of something else that would be much better?

You can also group actions such that you can do `dagger do ci someAction` vs `dagger do someAction` on the different environments. There's ways to reuse fields and reduce duplication, but it depends on how many actions and diverging fields you have.

**Reply by sagikazarmark** [reply]
I don't necessarily have a better idea (yet), I wanted to start a conversation about the topic and see what other people do.

Having access to client env vars from actions would be nice, but I also get why it's not implemented that way.

> You can also group actions

I thought about that, but it feels a bit weird. Ideally, I'd want my pipelines to be portable across environments and user env vars/parameters to configure pipelines. Having completely different groups is like having different pipelines.

BTW here is what I'm working on right now: https://github.com/sagikazarmark/dagger-go-cli/blob/main/dagger.cue
This is where this question came up.
